docs/winrt/WinRT SxS Components.md

```diff
@@ -1 +1,153 @@
+}
+```
+
+Version 2 of the same component will end up with a different implicit namespace in metadata:
+
+```
+namespace Widgets_1.Contoso.Widgets {
+        runtime_class Doodad;
+}
+```
+
+### ABI & Platform support
+At the ABI, types will self-identify using the component prefix namespace. For example, the IInspectable::GetRuntimeClassName method for Doodad will return the string "Widgets_1.Contoso.Widgets.Doodad". This ensure that metadata resolution algorithms and metadata-based marshalling support in all versions of Windows that support it will be able to uniquely correlate runtime classes with metadata definitions, without risk of type confusion.
+
+The component-qualified type name will also be used in calculations for pinterfaces GUID calculations. This ensures that marshaling and dynamic casting via QueryInterface will not result in type confusion.
+
+Component authors that use explicit GUIDs for interfaces are encouraged, but not required to generate new GUIDs if the binary layout of an interface changes across major versions. 
+
+Type registrations will include the component namespace element in the registered type name. This ensures that activtion of different versions is unambiguous, and that an application can include more than one version of the component. Call to RoGetActivationFactory or RoActivateInstance should use the component-qualified name>
+
+### The C++ Developer Experience
+Developers using older versions of C++ /WinRT will see the qualified namespace name. Developers using enlightened versions of C++ /WinRT will use the components as if the component namespace does not exist. This ensures that only minimal work is required to migrate from version 1 to version 2 of a component. Migration work can focus on breaking changes across versions, rather than mechanical changes required to update references to version-specific names.
+
+Open Design Issues: Should it be a goal to allow a single C++ /WinRT consumer to work with two versions of a component within the same compilation unit?
+
+### The C# Developer Experience
+Clients using .NET Core 3.1 and earlier, as well as .NET Framework 4 will be able to use components, but will refer to types using component-qualified names.
+
+Clients using the .NET 5 projection will not use component-qualified names. Instead, the projections will be differentiated by assembly name.
+
+#### Consuming
+
+Open Design Issue: How do the projection tools and build system handle uniquely named assemblies for different versions of the same component?
+
+#### Authoring
+
+Open Design Issue: How do the projection tools and build system handle uniquely named assemblies for different versions of the same component? Can the single source file authoring experience provide a high-quality experience for implementing components?
+
+### ABITool & MIDLRT header creation support
+and implicitly calculated interface names.
+
+### Design Patterns
+
+#### Cross-version interoperability
```

Comment:
Another strategy is to define a specific "component" that is about cross-component interop. Like:

```c#
[component_interop]
component WidgetInterop {
    namespace Contoso.Widgets.Interop {
        interface IDoodadInterop {
            Guid Id { get; };
        }
    }
}

[component_version(1)]
component Widgets {
    namespace Contoso.Widgets {
        runtimeclass Widget : Contoso.Widgets.Interop.IDoodadInterop {
            void DoThing();
        }
    }
}

[component_version(2)]
component Widgets {
    namespace Contoso.Widgets {
        runtimeclass Widget : Contoso.Widgets.Interop.IDoodadInterop {
            String GetCake(UInt32 minLayers);
        }
    }
}
```

The rule is that [component_interop] is effectively always version 1 and cannot declare a version 2 of itself.

If you want to prevent inter-component crosstalk, you could add a `[component_interop(WidgetInterop)]` to your `component Widgets` entries, specifically making just those types available.

Or the specific `runtimeclass` gets a new `[interop(WidgetInterop.Contoso.Widgets.IDoodadInterop)]` attribute such that the IDoodadInterop is not directly projected on the type, but projections know to inject this "cloaked" interface for casting purposes.

Reply:
avoiding duplicated metadata is still desirable. That will need to be handled with care in this situation. Thinking about this more, it may be more practical for a component author to deliver versioned components with a shared version-stable dependency. If "shared" portions are in defined from multiple components, we may need to teach some tools to de-dup and flag unexpected inconsistencies.

Reply:
I think these are workable suggestions but not necessary to achieve the key requirements this feature is trying to address. The requirement can be met using existing tech, but could be cleaner with better tooling support. Can we table this aspect and treat it as a separate proposal / discussion once this design is settled?